src/neo4clj/query_builder.clj

```diff
@@ -62,81 +62,139 @@
+                                                              rels
+                                                              false))))]
                  (when-not (empty? returns)
-                   (vector (str "RETURN " (str/join "," (get-return-ref-ids returns))))))))
+                   (vector (str "RETURN " (str/join ", " (get-return-ref-ids returns))))))))
+
+(defn- lookup-unmatched-nodes
+  "Takes a list of node entries, known reference ids and a collection of relations.
+  Returns a collection of node queries and where quiries for the missing node reference ids."
+  [node-entries known-ref-ids rels]
+  (loop [unmatched-nodes []
+         remaining-rels rels
+         known-refs known-ref-ids]
+    (if (empty? remaining-rels)
+      (vector (remove nil? (map first unmatched-nodes))
+              (remove nil? (map second unmatched-nodes)))
+      (let [{:keys [from to]} (first rels)]
+        (recur (conj unmatched-nodes
+                     (node-reference known-refs node-entries from true true)
+                     (node-reference known-refs node-entries to true true))
+               (rest remaining-rels)
+               (conj known-refs (node-ref-id from) (node-ref-id to)))))))
+
+(defn- lookup-non-existing-graph-rel
+  "Takes a relation lookup representation and generates a bolt query with a
+  where part to test that the given relationship doesn't exists"
+  ([{:keys [from to] :as rel}]
+   (str "NOT " (cypher/relationship
+                (str "(" (or (:ref-id from) from) ")")
+                (str "(" (or (:ref-id to) to) ")")
+                (dissoc rel :ref-id)))))
+
+(defn- lookup-graph-single-matches-and-wheres
+  "Takes a list of node entries, known reference ids, relations to
+  check for non-existence and a collection of where queries.
+  Returns a query with missing node matches, non existent relation
+  wheres and other wheres given."
+  [node-entries known-ref-ids not-exists-rels where-parts]
+  (let [[unmatched-node-lookups unmatched-node-wheres] (lookup-unmatched-nodes node-entries known-ref-ids not-exists-rels)]
+    (str (when (not (empty? unmatched-node-lookups))
+           (str " MATCH " (str/join " MATCH " unmatched-node-lookups)))
+         " WHERE "
+         (str/join " AND " (concat where-parts
+                                  unmatched-node-wheres
+                                  (map lookup-non-existing-graph-rel not-exists-rels))))))
+
+(defn lookup-graph-query
+  "Takes a list of node-entries and relations. Returns a query string
+  to lookup the nodes represented by the given relationships."
+  [node-entries rels]
+  (let [grouped-relations (group-by #(false? (:exists %)) rels)
+        exists-rels (get grouped-relations false)
+        not-exists-rels (get grouped-relations true)
+        [rel-queries where-parts known-ref-ids] (generate-relation-queries #{} node-entries exists-rels true)]
+      (str/trim (str (when (not (empty? rel-queries))
```

Comment:
Consider if the trimming overhead (human readability) is worth the performance impact.

Reply:
The cost is 0.1 microsecond in Criterium, so I think it is worth it